examples/discrete/discrete_dqn.py

```diff
@@ -25,12 +28,14 @@ def main():
     # you can define other net by following the API:
     # https://tianshou.readthedocs.io/en/master/tutorials/dqn.html#build-the-network
     env = gym.make(task, render_mode="human")
-    state_shape = env.observation_space.shape or env.observation_space.n
-    action_shape = env.action_space.shape or env.action_space.n
+    env.action_space = cast(gym.spaces.Discrete, env.action_space)
+    space_info = SpaceInfo.from_env(env)
+    state_shape = space_info.observation_info.obs_shape
+    action_shape = space_info.action_info.action_shape
     net = Net(state_shape=state_shape, action_shape=action_shape, hidden_sizes=[128, 128, 128])
     optim = torch.optim.Adam(net.parameters(), lr=lr)
 
-    policy = ts.policy.DQNPolicy(
+    policy: ts.policy.DQNPolicy = ts.policy.DQNPolicy(
```

Comment:
I also found it weird since there is no ambiguity of which policy to use, but mypy was throwing this error in this contexts:

```
examples/discrete/discrete_dqn.py:38: error: Need type annotation for "policy"  [var-annotated]
        policy = ts.policy.DQNPolicy(
                 ^
Found 1 error in 1 file (checked 1 source file)
```

I'll check in more depth why it complains.

Reply:
Resolved, cf. https://github.com/thu-ml/tianshou/pull/1077#discussion_r1546459488.